src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+{
+    public delegate void DeviceEventDelegate(DeviceInfo device);
+    public delegate void DeviceDataEventDelegate<DataT>(DeviceInfo device, DataT newState);
+
+    public class SoundManager : IMMNotificationClient
+    {
+        private IMMDeviceEnumerator _mmDeviceEnumerator;
+
+        private List<SoundClient> _clients = new();
+
+        public event DeviceEventDelegate DeviceRemoved;
+        public event DeviceEventDelegate DeviceAdded;
+        public event DeviceDataEventDelegate<PROPERTYKEY> DevicePropertyValueChanged;
+        public event DeviceDataEventDelegate<uint> DeviceChanged;
+
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        public SoundManager()
+        {
+            _mmDeviceEnumerator = Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid(ComCLSIDs.MMDeviceEnumeratorCLSID))) as IMMDeviceEnumerator;
+            if (_mmDeviceEnumerator == null)
+            {
+                throw new Exception("Creating device enumerator failed");
+            }
+
+            _mmDeviceEnumerator.RegisterEndpointNotificationCallback(this);
+        }
+
+        public SoundClient RegisterClientForDefault(DeviceType type, params DeviceRole[] roles)
+        {
+            var deviceId = GetDefaultDeviceId(type, roles);
+            var client = RegisterClient(deviceId, type, roles);
+
+            return client;
+        }
+
+        public SoundClient RegisterClientForDeviceId(string deviceId)
+        {
+            return RegisterClient(deviceId);
+        }
+
+        private SoundClient RegisterClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
+        {
+            var client = CreateClient(deviceId, type, roles);
+
+            //Map events
+            client.VolumeChanged += HandleVolumeChanged;
+            client.MuteChanged += HandleMuteChanged;
+
+            _clients.Add(client);
+
+            return client;
+        }
+
+        private void HandleMuteChanged(DeviceInfo device, bool newState)
+        {
+            MuteChanged?.Invoke(device, newState);
+        }
+
+        private void HandleVolumeChanged(DeviceInfo device, float newState)
+        {
+            VolumeChanged?.Invoke(device, newState);
+        }
+
+        public IEnumerable<SoundClient> GetClientsForDeviceId(string deviceId)
+        {
+            return _clients.Where(x => x.DeviceInfo.Id == deviceId);
+        }
+
+        public string GetDefaultPlaybackDeviceId(DeviceRole role = DeviceRole.Multimedia)
+        {
+            return GetDefaultDeviceId(DeviceType.Render, role);
+        }
+
+        public string GetDefaultRecordingDeviceId(DeviceRole role = DeviceRole.Communications)
+        {
+            return GetDefaultDeviceId(DeviceType.Capture, role);
+        }
+
+        public string GetDefaultDeviceId(DeviceType type, params DeviceRole[] roles)
+        {
+            if (_mmDeviceEnumerator == null)
```

Comment:
Can this condition ever occur in a working instance? If not, then we don't need this check.